File: 8/b.ts
// const input = await Bun.file("test.txt").text();
const input = await Bun.file("input.txt").text();

let res = 0;

const lines = input.split("\n");

const sequence = lines.shift()!;
lines.shift();

const obj: Record<
  string,
  {
    R: string;
    L: string;
  }
> = {};

lines.forEach((line) => {
  const [key, value] = line.split(" = ");
  obj[key] = {
    L: value.slice(1, 4),
    R: value.slice(6, 9),
  };
});

const gcd = (a: number, b: number): number => (b == 0 ? a : gcd(b, a % b));
const lcm = (a: number, b: number): number => (a / gcd(a, b)) * b;
const lcmAll = (ns: number[]) => ns.reduce(lcm, 1);

let i = 0;
let current = Object.keys(obj).filter((k) => k.endsWith("A"));
let prevCurrent: string[][] = [];
let loops = current.map(() => -1);
while (true) {
  if (i > sequence.length - 1) i = 0;
  let a = sequence[i];
  current = current.map((c) => obj[c][a as "R" | "L"]);
  for (let j = 0; j < prevCurrent.length; j++) {
    for (let k = 0; k < current.length; k++) {
      if (loops[k] !== -1 || !current[k].endsWith("Z")) continue;
      if (current[k] === prevCurrent[j][k]) {
        loops[k] = j + 1;
        console.log(j + 1, res);
      }
    }
  }
  prevCurrent.unshift(current);

  res++;
  i++;
  // console.log(prevCurrent, loops);
  if (loops.every((a) => a !== -1)) break;
  // await new Promise((r) => setTimeout(r, 1000));
  if (res % 1000000 === 0) console.log(res);
}

console.log(lcmAll(loops));

File: 8/b.go
package main

import (
	"fmt"
	"os"
	"strings"
)

type Node struct {
	right string
	left  string
}

func gcd(a, b int) int {
	for b == 0 {
		return a
	}
	return gcd(b, a%b)
}

func main() {
	file, _ := os.ReadFile("input.txt")

	lines := strings.Split(string(file), "\n")

	sequence := lines[0]
	nodesMap := make(map[string]Node)
	current := make([]string, 0)
	for i := 2; i < len(lines); i++ {
		key, rawValue, _ := strings.Cut(lines[i], " = ")
		if key[2] == 'A' {
			current = append(current, key)
		}
		nodesMap[key] = Node{left: rawValue[1:4], right: rawValue[6:9]}
	}

	currentLen := len(current)
	previous := make([][]string, 0)
	loops := make([]int, currentLen)

	for i := 0; ; i++ {
		newCurrent := make([]string, currentLen)
		for j := 0; j < currentLen; j++ {
			newCurrent[j] = nodesMap[current[j]].left
			if sequence[i%len(sequence)] == 'R' {
				newCurrent[j] = nodesMap[current[j]].right
			}
		}
		current = newCurrent
		previousLen := len(previous)
		for j := previousLen - 1; j >= 0; j-- {
			for k := 0; k < currentLen; k++ {
				if loops[k] == 0 &&
					current[k][2] == 'Z' &&
					previous[j][k] == current[k] {
					loops[k] = previousLen - j
				}
			}
		}
		isFinished := true
		for _, v := range loops {
			if v == 0 {
				isFinished = false
				break
			}
		}
		if isFinished {
			break
		}
		previous = append(previous, current)

	}
	res := 1
	for _, v := range loops {
		res = res * v / gcd(res, v)
	}
	fmt.Println(res)
}
